src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py

```diff
@@ -0,0 +1,225 @@
+        experiment_date = experiment.get("date")
+        # intermediate NWB files from each day will be saved in the nwb_output_folder
+        experiment_output_folder = nwb_output_folder / f"{experiment_date}"
+        experiment_output_folder.mkdir(parents=True, exist_ok=True)
+        print(f"\nExperiment Date: {experiment_date}-------------------\n")
+
+        # session-level information
+        sessions = experiment.get("sessions", [])
+        for session in sessions:
+            stimulus_name = session.get("stimulus_name")
+            stimulus_folder = Path(session.get("stimulus_folder"))
+            mworks_folder = Path(session.get("mworks_folder"))
+            data_folder = Path(session.get("data_folder"))
+            intan_file_path = data_folder / "info.rhd"
+            assert intan_file_path.exists(), f"{intan_file_path} does not exist"
+            notes = session.get("notes")
+
+            print(f"  Session Stimulus Name: {stimulus_name}")
+            print(f"  - Stimulus Folder: {stimulus_folder}")
+            print(f"  - MWorks Folder: {mworks_folder}")
+            print(f"  - Data Folder: {data_folder}")
+            print(f"  - Notes: {notes}\n")
+
+            # session info
+            session_metadata = {
+                "project_name": project,
+                "subject": subject,
+                "recording_id": data_folder.name,
+                "pipeline_version": pipeline_version,
+            }
+
+            # Extract stimulus/behavioral events from MWorks files
+            mworks_processed_file_path = parse_mworks_file(
+                mworks_folder, data_folder, mworks_folder
+            )  # save MWorks results in the same folder
+            print(f"  - MWorks processed file saved to: {mworks_processed_file_path}")
+
+            session_nwb_filepath = convert_session_to_nwb(
+                session_metadata=session_metadata,
+                intan_file_path=intan_file_path,
+                mworks_processed_file_path=mworks_processed_file_path,
+                stimuli_folder=stimulus_folder,
+                thresholindg_pipeline_kwargs=thresholindg_pipeline_kwargs,
+                psth_kwargs=psth_kwargs,
+                output_dir_path=experiment_output_folder,
+                stub_test=stub_test,
+                verbose=verbose,
+                add_thresholding_events=add_thresholding_events,
+                add_psth=add_psth,
+                add_stimuli_media_to_nwb=add_stimuli_media_to_nwb,
+                stimuli_are_video=stimuli_are_video,
+                ground_truth_time_column=ground_truth_time_column,
+                add_amplifier_data_to_nwb=add_amplifier_data_to_nwb,
+                add_psth_in_pipeline_format_to_nwb=add_psth_in_pipeline_format_to_nwb,
+            )
+
+            # add quality control if stimulus_name contains "normalizer"
+            if "normalizer" in stimulus_name:
+                with NWBHDF5IO(session_nwb_filepath, mode="r") as io:
```

Comment:
Suggestion:

Abastract this away into a function that takes an in-memory nwb and calculates outputs the data frame `calculate_quality_metrics_from_nwb(nwbfile)`

That way, the code flow would be simpler and you can modify the function nwihtout modifying the running script.

Reply:
Agreed and updated